Populate switch tweaks

CyrusNajmabadi opened the pull request:
This change tweaks 'Populate Switch' to provide from 1-3 fixes to the user depending on the state of the switch.  The potential fixes provided are:

1. Add missing cases.
2. Add missing 'default' case
3. Add both

CyrusNajmabadi commented:
Tagging @Hosch250

Hosch250 commented:
I just realized I missed a test case:

```
        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsPopulateSwitch)]
        public async Task EnumValueIsCastFromInteger()
        {
            await TestMissingAsync(
            @"
namespace ConsoleApplication1
{
    enum MyEnum
    {
        Fizz, Buzz, FizzBuzz
    }
    class MyClass
    {
        void Method()
        {
            var e = MyEnum.Fizz;
            switch ([|e|])
            {
                case (MyEnum)0:
                    break;
            }
        }
    }
}
");
        }
```

CyrusNajmabadi commented:
Yes.  I think the more appropriate thing ot do would be to see which enum *values* have been used in the switch, and to offer to add any members that have values that are not represented.  I will fix this up as well.

jmarolf commented:
:+1: do we have anyone who is willing to pull this down and take it for a spin?

CyrusNajmabadi commented:
Ok, this is ready for review.

Overall changes:

1. Moved to IOperation so we could have a single analyzer for C# and VB.
2. Moved to have a lot more code sharing in the code fix provider.
3. Removed logic around 'Flags' enums.  As this is now just a hidden analyzer, it's ok if we still offer the feature for those who want it on Flags enums.
4. Made it so that the user can choose between adding missing cases, adding hte default cause, or doing both.

CyrusNajmabadi commented:
retest vsi please

Hosch250 commented:
Have you added the test cases for when an integer is being cast to an enum type yet?

CyrusNajmabadi commented:
Not yet!  but i will add it.

Hosch250 commented:
This looks good to me.  I don't have time just now, but if this isn't handled fairly soon, I will take it for a stress test and see if I can't find a bug or two.

Hosch250 commented:
Interesting edge-case: When you type `switch (expr)` without braces, the quick-fix becomes available.  It adds the values correctly, but does not add the braces.  (C# version.)

Hosch250 commented:
Serious bug: It crashes with a non-fatal exception for the following code:
```
Dim expr = FileMode.CreateNew
Select Case expr
    Case 
End Select
```
```
Dim expr = FileMode.CreateNew
Select Case expr
    Case File    ' and any other incomplete expression
End Select
```
```
Dim expr = CreateNew    ' Imports System.IO.FileMode
Select Case expr
    Case Cre    ' and any other incomplete expression
End Select
```

Hosch250 commented:
> 
```
private static bool HasDefaultCase(ISwitchCase switchCase)
{
    foreach (var clause in switchCase.Clauses)
    {
        if (clause.CaseKind == CaseKind.Default)
        {
            return true;
        }
    }
    return false;
}
```

This can be optimized by iterating backwards.  In VB.NET, the default case is always the last one, and in C#, it will be the last one probably over 99% of the time.

Hosch250 commented on src/Features/Core/Portable/PopulateSwitch/PopulateSwitchHelpers.cs:
```diff
@@ -0,0 +1,131 @@
+
+        private static bool HasDefaultCase(ISwitchCase switchCase)
+        {
+            foreach (var clause in switchCase.Clauses)
```
This can be optimized by iterating backwards.  VB.NET requires the default case to be the last one, and C# will have it be the last one over 99% of the time.

Hosch250 commented on src/Features/Core/Portable/PopulateSwitch/PopulateSwitchHelpers.cs:
```diff
@@ -0,0 +1,131 @@
+
+        public static bool HasDefaultCase(ISwitchStatement switchStatement)
+        {
+            foreach (var switchCase in switchStatement.Cases)
```
This can be optimized by iterating backwards.  VB.NET requires the default case to be the last one, and C# will have it be the last one over 99% of the time.

Hosch250 commented on src/Features/Core/Portable/PopulateSwitch/PopulateSwitchDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,74 @@
+            out bool missingCases, out bool missingDefaultCase)
+        {
+            missingDefaultCase = !PopulateSwitchHelpers.HasDefaultCase(switchStatement);
+            missingCases = false;
```
Assignment never used.

Hosch250 commented on src/Features/Core/Portable/PopulateSwitch/PopulateSwitchDiagnosticAnalyzer.cs:
```diff
@@ -0,0 +1,74 @@
+            missingCases = false;
+
+            var switchExpression = switchStatement.Value;
+            var switchExpressionType = switchExpression?.Type;
```
Variable never used.

Hosch250 commented on src/Features/Core/Portable/PopulateSwitch/PopulateSwitchCodeFixProvider.cs:
```diff
@@ -0,0 +1,142 @@
+            var switchStatement = (ISwitchStatement)model.GetOperation(switchNode, cancellationToken);
+            var enumType = switchStatement.Value.Type;
+
+            var containsDefaultCase = PopulateSwitchHelpers.HasDefaultCase(switchStatement);
```
Variable never used.

Hosch250 commented:
Finally, I don't see a full test suite for casting integer values to enum values.  There should be tests for when some cases exist as casts and the fix does not add their values in both normal and `using static` references.  There should be a test that the warning doesn't show when all the cases exist as casts.  And the VB.NET suite doesn't check this at all (I don't think it needs to `DirectCast`, I think it just works as an integer--double check).

CyrusNajmabadi commented:
With testing one of the things to avoid is actually dependent APIs that the feature actually depends on.  In this case (no pun intended) there is the expectation that hte lower level API properly return constant values.  As such, it suffices to do a quick test to validate that that scenario works, without having to exhaustively test every permutation.

Hosch250 commented:
OK.  I'll look this over again later.

Hosch250 commented:
Looks good.

CyrusNajmabadi commented:
Thanks @Hosch250!

rchande commented on src/Features/Core/Portable/PopulateSwitch/PopulateSwitchHelpers.cs:
```diff
@@ -0,0 +1,131 @@
+            return false;
+        }
+
+        public static ICollection<ISymbol> GetMissingEnumMembers(ISwitchStatement switchStatement)
```
Would be really impressive if this supported the "completionlist tag" feature (where the IDE will act like static members of a non-enum type are enum members).

rchande commented on src/Workspaces/CSharp/Portable/CodeGeneration/CSharpSyntaxGenerator.cs:
```diff
@@ -2510,6 +2510,66 @@ public override SyntaxNode InsertParameters(SyntaxNode declaration, int index, I
 
+        public override IReadOnlyList<SyntaxNode> GetSwitchSections(SyntaxNode switchStatement)
+        {
+            var statement = switchStatement as SwitchStatementSyntax;
```
What can this get called against that isn't a SwitchStatementSyntax?

rchande replied:
Oh, this is the generator, not your analyzer, nvm.

davkean commented:
Why's this in master? IOperation is turned off, right?

Hosch250 commented:
@davkean It worked when I pulled, built, and ran it.

CyrusNajmabadi commented:
Closing in favor of https://github.com/dotnet/roslyn/pull/10846
